compiler/src/iree/compiler/GlobalOptimization/DemoteContractionInputsToBF16.cpp

```diff
@@ -0,0 +1,122 @@
+    for (auto operand : linalgOp->getOperands()) {
+      auto operandType = dyn_cast<RankedTensorType>(operand.getType());
+      if (!operandType ||
+          operandType.getElementType() != rewriter.getF32Type()) {
+        return failure();
+      }
+    }
+    Location loc = linalgOp.getLoc();
+    SmallVector<Value> demotedInputs;
+    for (auto inputOperand : linalgOp.getDpsInputOperands()) {
+      auto input = inputOperand->get();
+      auto inputType = cast<RankedTensorType>(input.getType());
+      auto demotedInputType =
+          RankedTensorType::get(inputType.getShape(), rewriter.getBF16Type(),
+                                inputType.getEncoding());
+      SmallVector<AffineMap> maps(
+          2, AffineMap::getMultiDimIdentityMap(inputType.getRank(),
+                                               rewriter.getContext()));
+      SmallVector<utils::IteratorType> iteratorTypes(
+          inputType.getRank(), utils::IteratorType::parallel);
+      SmallVector<OpFoldResult> mixedSizes =
+          tensor::getMixedSizes(rewriter, loc, input);
+      Value empty = rewriter.create<tensor::EmptyOp>(loc, mixedSizes,
+                                                     rewriter.getBF16Type());
+      demotedInputs.push_back(
+          rewriter
+              .create<linalg::GenericOp>(
+                  loc, TypeRange{demotedInputType}, ValueRange{input},
+                  ValueRange{empty}, maps, iteratorTypes,
+                  [&](OpBuilder &b, Location loc, ValueRange args) {
+                    Value result = b.create<arith::TruncFOp>(
+                        loc, rewriter.getBF16Type(), args[0]);
+                    b.create<linalg::YieldOp>(loc, result);
+                  })
+              ->getResults()[0]);
+    }
+
+    if (isa<linalg::MatmulOp>(linalgOp)) {
+      rewriteOpWithNewInputs<linalg::MatmulOp>(rewriter, linalgOp,
+                                               demotedInputs);
+    } else if (isa<linalg::MatvecOp>(linalgOp)) {
+      rewriteOpWithNewInputs<linalg::MatvecOp>(rewriter, linalgOp,
+                                               demotedInputs);
+    } else if (isa<linalg::VecmatOp>(linalgOp)) {
+      rewriteOpWithNewInputs<linalg::VecmatOp>(rewriter, linalgOp,
+                                               demotedInputs);
+    } else if (isa<linalg::BatchMatmulOp>(linalgOp)) {
+      rewriteOpWithNewInputs<linalg::BatchMatmulOp>(rewriter, linalgOp,
+                                                    demotedInputs);
+    } else if (isa<linalg::BatchMatvecOp>(linalgOp)) {
+      rewriteOpWithNewInputs<linalg::BatchMatvecOp>(rewriter, linalgOp,
+                                                    demotedInputs);
+    } else if (isa<linalg::BatchVecmatOp>(linalgOp)) {
+      rewriteOpWithNewInputs<linalg::BatchVecmatOp>(rewriter, linalgOp,
+                                                    demotedInputs);
+    } else {
+      return failure();
+    }
```

Comment:
We can rewrite it with `mlir::clone`: https://github.com/llvm/llvm-project/blob/b33a131c828437efff36179458562e62774da881/mlir/include/mlir/Dialect/Utils/StructuredOpsUtils.h#L137-L141

I.e.,

```
clone(rewriter, linalgOp, op.getResultTypes(), demotedInputs);
```

(note: I'm not sure if `op.getResultTypes()` works. But there are equivalent methods somewhere.)

Reply:
@hanhanW Interesting! I tried, but this fails with the following error message. If I understand correctly, this is saying that linalg named ops (like here `batch_vecmat`) already have a `bb` region like `linalg.generic` ops, and when we apply `clone` to them, it clones that, so using `clone` to change inputs element type fails as it would need to perform a rewrite of the region itself.

```
error: 'linalg.batch_vecmat' op expected type of bb argument #0 ('f32') to match element or self type of the corresponding operand ('bf16')
  %0 = linalg.batch_vecmat ins(%arg0, %arg1 : tensor<4x250xf32>, tensor<4x250x500xf32>)
       ^
within split at /home/benoit/iree/compiler/src/iree/compiler/GlobalOptimization/test/demote_contraction_inputs_to_bf16.mlir:91 offset :5:8: note: see current operation: 
%4 = "linalg.batch_vecmat"(%1, %3, %arg2) <{operandSegmentSizes = array<i32: 2, 1>}> ({
^bb0(%arg3: f32, %arg4: f32, %arg5: f32):
  %5 = "arith.mulf"(%arg3, %arg4) <{fastmath = #arith.fastmath<none>}> : (f32, f32) -> f32
  %6 = "arith.addf"(%arg5, %5) <{fastmath = #arith.fastmath<none>}> : (f32, f32) -> f32
  "linalg.yield"(%6) : (f32) -> ()
}) {linalg.memoized_indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d2)>, affine_map<(d0, d1, d2) -> (d0, d2, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>]} : (tensor<4x250xbf16>, tensor<4x250x500xbf16>, tensor<4x500xf32>) -> tensor<4x500xf32>
```

Reply:
oh, I think you are right. The block arguments are not updated. Sorry for the bad suggestion.. Let's keep it to what you have.

Reply:
I think the method does not work well when types change.

Reply:
[optional] Below was written by me a couple years ago. With the trick, you dont need to pass `linalgOp` and `demotedInputs` for all the cases. Lambda with template is only available after c++20, so I used `std::remove_pointer_t<decltype(typePtr)>` to workaround it. We can probably do something better after c++17, but I don't follow that for a while. Someone removed my TODO comments, and I can't find it atm.

https://github.com/openxla/iree/blob/73f5947ff170fbf1ea8e415c019673cb7dd06aea/compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/StableHLOToLinalgReduce.cpp#L666-L682

Reply:
ah thanks - sorry, already merged. but, nice trick!